Created FAB and click on Team List Pokemon

mwyoung opened the pull request:
I have created the FAB button, opening pokemon from this, adding details, a artwork fallback url (which has glide within glide), and a onclick listener for the teams. 
Let me know if I need to change anything on this.

mwyoung committed:
Tabs work, nothing else does

mwyoung committed:
Can now launch empty activity with back button

mwyoung committed:
Added FAB button

mwyoung committed:
Can now open pokemon list from plus button

mwyoung committed:
Changed title

mwyoung committed:
Now fills in details

mwyoung committed:
Merge branch 'master' into Fab

# Conflicts:
#	app/src/main/java/rec/games/pokemon/teambuilder/TeamListFragment.java
#	app/src/main/res/layout/team_list.xml

mwyoung committed:
Added fallback URL

mwyoung committed:
Now accuratly passes data to PokemonItemDetailActivity

mwyoung committed:
Can not click on team list member to get more information

mwyoung committed:
Fixed error on pokemon > 1000

hammockt commented:
Pretty good start, but I think PokemonListFragment and PokemonListAdapater might be problematic. My pull request #10 that introduces LiveDataLists is probably the preferred solution for those files and either you or I will have to merge that into those files. I personally think that I should merge first, but whatever you prefer is fine with me

mwyoung commented:
I am fine waiting for #10 to be completed. I can now pass around IDs, so for those I would just have to update to your current solution (which was my plan).

hammockt commented:
Go wild

mwyoung committed:
Started preferences
(No changes so far)

mwyoung committed:
Merge branch 'master' into Fab

# Conflicts:
#	app/src/main/java/rec/games/pokemon/teambuilder/MainActivity.java
#	app/src/main/java/rec/games/pokemon/teambuilder/PokemonListAdapter.java

mwyoung committed:
Merged changes from master

hammockt reviewed (approved):
So much better. LGTM

mwyoung committed:
Can now hide images

mwyoung committed:
Can now hide images in all locations

mwyoung commented:
I finally figured out how to get preferences in the activities, so that now works. The code does not look pretty, but it does work.

hammockt commented:
It isn't that bad

mwyoung commented:
If you are OK with it (along with Andrew), then I am fine merging it in.

mwyoung committed:
Added FAB to Pokemon list, FAB hides when scrolling

mwyoung commented:
I now feel pretty confident I am done with adding FABs. I just added one to the Pokemon list for searching as well. @magical, what do you think of this?
Edit: Added FAB to item selection

mwyoung committed:
Update JDK

mwyoung committed:
Item FAB now switches

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/PokemonItemDetailActivity.java:
```diff
@@ -67,12 +133,28 @@ public void sharePokeDetails(){
 		}
 	}
+
+	public void ifItemAddedFAB(){
```
The name of this method doesn't seem very clear to me.

mwyoung replied:
I can update the name. It was mostly to see if the item was actually added.

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/PokemonListAdapter.java:
```diff
@@ -35,6 +47,11 @@ public void onItemChanged(@Nullable Pokemon pokemon, int index)
 
+	public interface OnPokemonClickListener
+	{
+		void onPokemonClicked(int position);
```
The parameter here is the id, not the list position, right?

mwyoung replied:
It would be the ID. I will change this.

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/PokemonListAdapter.java:
```diff
@@ -58,49 +82,66 @@ public int getItemCount()
+		if(position > 0 && mPokemon != null)
+		{
+			Log.d(TAG, Integer.toString(mPokemon.getValue(position - 1).getId()));
+			return mPokemon.getValue(position - 1).getId(); //mPokemon ids start at 1
```
How do we know getValue won't return null here? I guess i'm confused about what data we know will be loaded at what time.

hammockt replied:
With how I initially setup the ViewModel and LiveDataList, the list will never have nulls inside. But that doesn't stop one from using a bad index to get a null value or an exception.

At this point in time in the code, we can guarantee that the Pokemon is a instance of DeferredPokemonResource, but it may not be a PokemonResource. We would need to check for that if you want to use anything special from the non-deferred class

hammockt replied:
But you bring up a pretty good point of whether or not we should be assuming the LiveData objects contain null values

magical replied:
Okay, that makes sense.

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/PokemonListFragment.java:
```diff
@@ -119,21 +112,79 @@ public void onClick(View v)
+
 		Intent intent = new Intent(getContext(), PokemonItemDetailActivity.class);
-		intent.putExtra(PokeAPIUtils.POKE_ITEM, position); //temporary assignment
+		intent.putExtra(PokeAPIUtils.POKE_ITEM, pokemonID+1); //assign id
```
Why plus one?

mwyoung replied:
I believe there is a +1 there as the ID was off by 1. I am going to change how it is passed back and have the +1 there instead of in this function.

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/TeamAdapter.java:
```diff
@@ -26,52 +47,88 @@ public int getItemCount() {
         private ImageView image;
         private TextView name;
-        private Object listener; // TODO
+        //private Object mListener;
```
Delete this line.

mwyoung committed:
Can pass team to add to

mwyoung committed:
Passes team name to pokemon view, knows if adding to team

mwyoung committed:
Save user text

mwyoung commented:
Passing around the TEAM_MEMBER is very iffy, and probably should be changed. I just didn't know where to put it. 
I also got it to successfully work, and shows if the pokemon is added or not.

hammockt commented:
For TEAM_MEMBER, could you define it in strings.xml for now? Hardcoding it all over the place will make it fun to refactor later down the road

magical reviewed (commented):
Just to check, is this the workflow for adding a pokemon to a team?

    Team page, click plus button ->
    Pokemon list activity, click pokemon ->
    Pokemon detail activity, click plus button ->
    plus button changes to already-in-team button and user remains on detail activity

I was picturing a slightly shorter flow

    Team page, click plus button ->
    Pokemon list activity, click pokemon ->
    returns to team page with the clicked pokemon added

mwyoung committed:
Changed TEAM_MEMBER to string.xml entry

mwyoung commented:
Done.

magical commented:
> 
> 
> For TEAM_MEMBER, could you define it in strings.xml for now? Hardcoding it all over the place will make it fun to refactor later down the road

I'd make it a constant on the Team class.

```
class Team {
    public final static String EXTRA_TAG = Team.class.getName()
    ....
}

```

mwyoung committed:
Update from feedback

mwyoung commented:
@magical, in terms of the flow I could do it the second way. I was just thinking that when using the first it would allow the user to see which Pokemon is already on the team. This would then mean there could be no duplicates as well (which we might or might not want).
If we go directly from the list, that means that the Pokemon details could not be seen to see if the user actually wanted that Pokemon or if more details are needed. It would also mean that the current position in the list would not be saved as well if getting multiple Pokemon.

magical commented on app/src/main/java/rec/games/pokemon/teambuilder/PokemonItemDetailActivity.java:
```diff
@@ -153,7 +152,7 @@ public void shareToBrowser(){
 	}
 
-	public void ifItemAddedFAB(){
+	public void itemFAB(){
```
Now it's even less clear what this method does. Please be descriptive, and use verbs. Something like toggleFABIcon, or addOrRemovePokemonFromTeam.

mwyoung replied:
Sorry about that. I have just updated using the second suggestion.

magical replied:
Thanks. Much better :)

mwyoung committed:
Slight naming updates

magical commented:
@mwyoung Good point about duplicates. I'm not particularly attached to either approach, i just wanted to make sure i'm on the same page as you.

mwyoung commented:
If you both are OK with my code then I am ready to merge. If you also find you need to change something after merging feel free to do so.
